### PR TITLE
merge feature/rev3 complete

### LINKED ISSUE #24: add CAN frame on VCU to show current pedal thresholds, and current "calculated torque" from pedal position
this issue should be closed before #23
for torque, send "torque1" and "torque2" from verify_pedals method: 
https://github.com/KSU-MS/KS6e-VCU-firmware/blob/main/src/pedal_handler.cpp#L131-L138
as well as other values of interest
for thresholds, need to decide on periodic sending or triggered sending (ie. query by CAN tool)
- apps1 lower bound
- apps1 torque calc start
- apps1 torque calc end
- apps1 upper bound
- apps2 lower bound
- apps2 torque calc start
- apps2 torque calc end
- apps2 upper bound
- _BSE lower bound_
- BSE brake active threshold
- _BSE upper bound_
things in italics dont exist yet but they should

### LINKED ISSUE #26: VCU acts as though all dash buttons are pressed by default if no "buttons" can frame (ID 0x0eb) has been received
because the button statuses class initialized values = 0, vcu will toggle torque modes as if a button were depressed
for any other button interactions, the VCU will think that the buttons are all pressed due to the current logic
options: 
- [ ] Implement a timeout check where button stats gets locked out if dash messages has not been received
- [x] change button logic ("pressed" logic) to not inverter- requires dash to invert its output

### LINKED ISSUE #28: refactor calculate_torque
currently the calculate torque function is written pretty very poorly, for example some of the variables that will set the torque to _**ZERO**_ are not accessible by anything else because they only exist in the function. refactor it and the pedals class as necessary to improve readability/code quality/“visibility” of what the function does

### LINKED ISSUE #31: run verify_pedals in all states so that pedal problems can be detected when not in ready to drive

### PR DESCRIPTION
**Pull request by mathbrook**

### CODE REVIEW

**Comment by mathbrook** [comment]
need to work on #24 
and #23

File: src/main.cpp [hunk 1]
```diff
@@ -15,13 +15,15 @@
...
+#include "device_status.h"
+
+#define NEBUG
 
 // Metro timers for inverter:
 Metro timer_mc_kick_timer = Metro(50, 1);
 Metro timer_inverter_enable = Metro(2000, 1); // Timeout failed inverter enable
 Metro timer_motor_controller_send = Metro(100, 1);
-Metro timer_current_limit_send = Metro(500,1);
+Metro timer_current_limit_send = Metro(500, 1);
```
Comment (mathbrook) [inline src/main.cpp]: probably need to increase this to be more responsive when driving on track
also need a disable flag for the current limting

File: src/pedal_handler.cpp [hunk 2]
```diff
@@ -44,97 +57,113 @@ int PedalHandler::calculate_torque(int16_t &motor_speed, int &max_torque, bool r
...
-        WriteToDaqCAN(tx_msg2);
-    }
     // only uses front brake pedal
     brake_is_active_ = (brake1_ >= BRAKE_ACTIVE);
     return brake_is_active_;
 }
 
+void PedalHandler::send_readings()
+{
+    if (pedal_out->check())
```
Comment (mathbrook) [inline src/pedal_handler.cpp]: ask stakeholders what frequency they want this to be reported

File: src/pedal_handler.cpp [hunk 3]
```diff
@@ -161,12 +190,12 @@ void PedalHandler::verify_pedals(
         // mcu_status.set_no_accel_implausability(true);
         accel_is_plausible = true;
     }
-    Serial.printf("Torque 1: %d Torque 2: %d Torque Sum: %f Torque Average %f Torque Difference: %f\n",torque1,torque2,torqSum,torqAvg,torqDiff);
+    Serial.printf("Torque 1: %d Torque 2: %d Torque Sum: %f Torque Average %f Torque Difference: %f\n", torque1, torque2, torqSum, torqAvg, torqDiff);
```
Comment (mathbrook) [inline src/pedal_handler.cpp]: these need to be in a struct or something so we can send over CAN to monitor remotely

File: include/KS2eVCUgpios.hpp
Comment (mathbrook) [inline include/KS2eVCUgpios.hpp:1]: include old VCU pindefs but have a flag for which to compile

File: src/pedal_handler.cpp (hunk 2 above)
Comment (jstri114) [reply src/pedal_handler.cpp]: off the hip 100ms? idk tho

File: src/main.cpp (hunk 1 above)
Comment (jstri114) [reply src/main.cpp]: agreed

File: src/main.cpp [hunk 4]
```diff
@@ -104,6 +122,11 @@ void setup()
 
 void loop()
 {
+    // delay(10);
+    // uint16_t value = generateSineValue(millis(),1500/2,0.5,0,0);
+    // pedals.read_pedal_values_debug(value);
+    // pedals.run_pedals();
+    // pedals.calculate_torque(dummy_motor_speed,dummy_max_torque);
```
Comment (jstri114) [inline src/main.cpp:131]: so this would let you test the calculate torque functions and such without having a running car?

Comment (mathbrook) [reply src/main.cpp:131]: yeah this is what I used to test the pedalSensor class and refactored calculate_torque

File: src/ADC_SPI.cpp [hunk 5]
```diff
@@ -12,7 +12,7 @@
  */
 ADC_SPI::ADC_SPI()
 {
-	init(DEFAULT_SPI_CS, DEFAULT_SPI_SPEED);
+	// init(DEFAULT_SPI_CS, DEFAULT_SPI_SPEED);
 }
```
Comment (mathbrook) [inline src/ADC_SPI.cpp:16]: running "init" in the constructor would crash teensy b/c SPI.begin() gets called in the pedalHandler constructor and not in a function. this works

File: include/KS2eCAN.hpp [hunk 6]
```diff
@@ -27,12 +27,14 @@
...
+#define ID_VCU_BOARD_ANALOG_READS_ONE                               0xC9
+#define ID_VCU_BOARD_ANALOG_READS_TWO                               0xCA
 #define ID_DASH_BUTTONS                                             0xEB
-#define ID_DASH_BUSVOLT                                             0xD4
 #define ID_BMS_SOC                                                  0x6B3 //made this real!
 
 
-#define ID_BMS_INFO     0x6B1
+#define ID_BMS_CURRENT_LIMIT_INFO     0x6B1
+#define ID_BMS_PACK_VOLTAGE_INFO      0x6B2
```
Comment (mathbrook) [inline include/KS2eCAN.hpp:40]: add 0x6b2 message ID and made the names more verbose so its more clear what they are

**Comment by mathbrook** [comment]
- ran on dyno 3-2-2024
- EV active apps check passed (sensors unplugged, no spin, brake pressed, no spin, etc.)

File: src/pedal_handler.cpp [hunk 7]
```diff
@@ -1,27 +1,40 @@
...
+Metro debugPrint = Metro(100);
 
 // initializes pedal's ADC
 void PedalHandler::init_pedal_handler()
 {
-    pedal_ADC = ADC_SPI(DEFAULT_SPI_CS, DEFAULT_SPI_SPEED);
+    // TODO don't htink you actually have to init analog GPIOs on teensy
+    // for(int i; i < sizeof(analog_init_list)/sizeof(int);i++){
+    //     pinMode(analog_init_list[i],INPUT);
+    // }
```
Comment (mathbrook) [inline src/pedal_handler.cpp:11]: you in fact do not have to init them to do analogReads

File: include/KS2eCAN.hpp (hunk 6 above)
Comment (mathbrook) [reply include/KS2eCAN.hpp:40]: added

**Pull request merged**